examples/knative/templates/server.ps1

```diff
@@ -1,141 +1,156 @@
-            $headers = @{}
-            for($i =0; $i -lt $context.Request.Headers.Count; $i++) {
-               $headers[$context.Request.Headers.GetKey($i)] = $context.Request.Headers.GetValues($i)
+    param($url, $serverStopMessage)
+
+    Import-Module 'Microsoft.PowerShell.Utility'
+    Import-Module CloudEvents.Sdk
+
+    . ./handler.ps1
+
+    function Start-HttpCloudEventListener {
+        <#
+            .DESCRIPTION
+            Starts a HTTP Listener on specified Url
+        #>
+
+        [CmdletBinding()]
+        param(
+            [Parameter(
+                Mandatory = $true,
+                ValueFromPipeline = $false,
+                ValueFromPipelineByPropertyName = $false)]
+            [ValidateNotNull()]
+            [string]
+            $Url
+        )
+
+        $listener = New-Object -Type 'System.Net.HttpListener'
+        $listener.AuthenticationSchemes = [System.Net.AuthenticationSchemes]::Anonymous
+        $listener.Prefixes.Add($Url)
+
+        $cloudEvent = $null
+
+        try {
+            $listener.Start()
+
+            while ($true) {
+                $context = $listener.GetContext()
+
+                try {
+                    # Read Input Stream
+                    $buffer = New-Object 'byte[]' -ArgumentList 1024
+                    $ms = New-Object 'IO.MemoryStream'
+                    $read = 0
+                    while (($read = $context.Request.InputStream.Read($buffer, 0, 1024)) -gt 0) {
+                        $ms.Write($buffer, 0, $read);
+                    }
+                    $bodyData = $ms.ToArray()
+                    $ms.Dispose()
+
+                    # Read Headers
+                    $headers = @{}
+                    for ($i = 0; $i -lt $context.Request.Headers.Count; $i++) {
+                        $headers[$context.Request.Headers.GetKey($i)] = $context.Request.Headers.GetValues($i)
+                    }
+
+                    $cloudEvent = ConvertFrom-HttpMessage -Headers $headers -Body $bodyData
+
+                    if ([System.Text.Encoding]::UTF8.GetString($bodyData) -eq $serverStopMessage) {
+                        # Server Stop request
+                        $context.Response.StatusCode = [int]([System.Net.HttpStatusCode]::OK)
+                        $context.Response.Close();
+                        # Set function reault
+                        $true
+                        break
+                    }
+
+                    if ( $cloudEvent -ne $null ) {
+                        try {
+                            Process-Handler -CloudEvent $cloudEvent | Out-Null
+                            $context.Response.StatusCode = [int]([System.Net.HttpStatusCode]::OK)
+                        }
+                        catch {
+                            Write-Error "$(Get-Date) - Handler Processing Error: $($_.Exception.ToString())"
+                            $context.Response.StatusCode = [int]([System.Net.HttpStatusCode]::InternalServerError)
```

Comment:
For now this is totally fine. We can think about improving this to distinct between BAD_REQUEST (4xx, e.g. data payload) error types and INTERNAL_SERVER_ERROR (5xx) so the response of the context would be a 4xx or 5xx depending on the handler exception.

Reply:
The benefit of having distinct exceptions is that context response would be understood by sender, e.g. broker which then would not retry if it's 4xx

Reply:
agree

Reply:
SGTM. Can you please add a `TODO:` at this section so we don't forget? Currently this might lead to unnecessary retries from brokers even in case of unrecoverable errors (e.g. 4xx).

Reply:
added TODOs for HTTP status code